### PR TITLE
Add decoders for heterogenenous arrays and maps.

### PR DESCRIPTION
**Pull request by KtorZ**

### CODE REVIEW

**Comment by simonh1000** [comment]
This is super sweet!  As a matter of interest, why did you use a record to decode a heterogeneous array. I think I can see that such arrays are likely to have fixed positions, but an Elm alternative could have been a list of some custom type that represented the different types that could exist in the list?

**Comment by KtorZ** [comment]
I used a record because an heterogeneous array of elements is basically a product type, which is better modeled as a record IMO (nice accessors and structure).

An alternative could have been to use a 4-tuple (which really is the poor-man record ^^) or, as you indicate, a list of sum-type. The latter is relevant if you intend to process the array as an homogeneous list and do similar operation on each element. In this particular case, I don't think it'd a wise choice since each element in the array has its own semantic. Really, it's a record 😁

**Comment by simonh1000** [comment]
Yes, here it is a record - I was just thinking about other cases

File: src/Cbor/Decode.elm [hunk 1]
```diff
@@ -350,6 +368,13 @@ dict key value =
                 unsigned a |> D.andThen (\n -> D.loop ( n, [] ) finite)
 
 
+{-| Like 'array', but for heterogeneous dict.
+-}
+record : Decoder a -> Decoder a
+record =
+    Decoder (MajorType 5) << always << runDecoder
```
Comment (simonh1000) [inline src/Cbor/Decode.elm:393]: Does this work? Don't you need some code to either drop the keys as you run into them, or to take a decoder for the keys and return a Dict?

Comment (KtorZ) [reply src/Cbor/Decode.elm:393]: It works similarly to the `array` one, so `map` primitives are friends here. The decoding of key/value is to be done manually in that case. A `Dict` cannot be used for this is to cope precisely with heterogeneous structures which won't work with Elm's `Dict comparable a` type. This construct would also allow decoding records for which keys are not necessarily comparable, and heterogeneous too. For example, imagine a CBOR-map with 2 elements, the first one is a key:value from bool to int, and the second one a key:value from string to a list of bool. You would decode it as such:

```elm
type alias MyRecord = 
    { first : (Bool, Int)
    , second : (String, List Bool)
    }

D.decode <| D.record <|
    D.map2 MyRecord
        (D.pair D.bool D.int)
        (D.pair D.string (D.list D.bool))
````

We resort to representing each key:value pair as a 2-tuple.

Comment (simonh1000) [reply src/Cbor/Decode.elm:393]: cool - thx

**Pull request merged**